pkg/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,194 @@
+package telemetry
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/google/uuid"
+	"github.com/square/spiffe-aws-assume-role/internal/mocks"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+var anyLabels string = mock.Anything
+
+func TestInstrumentCalls(t *testing.T) {
+	metricSink := mocks.MetricSink{}
+	allowAllCalls(&metricSink)
+
+	telemetry, err := NewTelemetryForSink(&metricSink)
+	require.NoError(t, err)
+
+	emitMetrics := telemetry.Instrument([]string{"foo", "bar"}, &err)
+	emitMetrics()
+
+	metricSink.AssertCalled(t, "IncrCounterWithLabels",
+		[]string{"spiffe-aws-assume-role", "foo", "bar", "Calls"},
+		float32(1),
+		anyLabels)
+}
+
+func TestInstrumentLatency(t *testing.T) {
+	metricSink := mocks.MetricSink{}
+	allowAllCalls(&metricSink)
+
+	telemetry, err := NewTelemetryForSink(&metricSink)
+	require.NoError(t, err)
+
+	emitMetrics := telemetry.Instrument([]string{"foo", "bar"}, &err)
+	time.Sleep(time.Second * 1)
+	emitMetrics()
+
+	metricSink.AssertCalled(t, "SetGaugeWithLabels",
+		[]string{"spiffe-aws-assume-role", "foo", "bar", "Latency"},
+		mock.MatchedBy(greaterThanOrEqualFloat32(1000)),
+		anyLabels)
+}
+
+func TestInstrumentSuccess(t *testing.T) {
+	metricSink := mocks.MetricSink{}
+	allowAllCalls(&metricSink)
+
+	telemetry, err := NewTelemetryForSink(&metricSink)
+	require.NoError(t, err)
+
+	require.NoError(t, methodThatSucceeds(telemetry))
+
+	metricSink.AssertCalled(t, "IncrCounterWithLabels",
+		[]string{"spiffe-aws-assume-role", "foo", "bar", "Success"},
+		float32(1),
+		anyLabels)
+	metricSink.AssertCalled(t, "IncrCounterWithLabels",
+		[]string{"spiffe-aws-assume-role", "foo", "bar", "Failure"},
+		float32(0),
+		anyLabels)
+}
+
+func TestInstrumentFailure(t *testing.T) {
+	metricSink := mocks.MetricSink{}
+	allowAllCalls(&metricSink)
+
+	telemetry, err := NewTelemetryForSink(&metricSink)
+	require.NoError(t, err)
+
+	require.Error(t, methodThatFails(telemetry))
+
+	metricSink.AssertCalled(t, "IncrCounterWithLabels",
+		[]string{"spiffe-aws-assume-role", "foo", "bar", "Failure"},
+		float32(1),
+		anyLabels)
+	metricSink.AssertCalled(t, "IncrCounterWithLabels",
+		[]string{"spiffe-aws-assume-role", "foo", "bar", "Success"},
+		float32(0),
+		anyLabels)
+}
+
+func TestInstrumentHostname(t *testing.T) {
+	metricSink := mocks.MetricSink{}
+	allowAllCalls(&metricSink)
+
+	hostname := uuid.New().String()
+
+	telemetry, err := NewTelemetryForSinkAndHostname(&metricSink, hostname)
+	require.NoError(t, err)
+
+	telemetry.Instrument(nil, &err)()
+
+	hostnameLabel := []metrics.Label{{
+		Name:  "hostname",
+		Value: hostname,
+	}}
+
+	metricSink.AssertCalled(t, "IncrCounterWithLabels",
+		[]string{"spiffe-aws-assume-role", "Calls"},
```

Comment:
Checking to see if I understand how `copyAndAppend` works: if we were to call `Instrument` twice, would there be two "Calls" strings in the slice for the second `IncrCounterWithLabels` call?

Reply:
Actually hmmm, seems like `keys` itself (aka the source) isn't updated when calling `copyAndAppend`, so I'm doubting that this will happen.

Reply:
Yeah, that's actually the whole reason I wrote it: because the full metric name is an array/slice and we want to append different suffixes (Success, Failure, Calls, etc.) to the same "prefix" I wanted a way to do that without having the calls interfere with each other. As mentioned above I'm no Go wizard so maybe there's a more idiomatic way to accomplish this.

Reply:
Looks like copying and appending is the safest way to prevent the calls from interfering with each other (see url I linked in another comment about unexpected `append` behavior).